flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/utils/AtomicRtasITCaseBase.java

```diff
@@ -0,0 +1,223 @@
+            tEnv.executeSql("create table " + tableName + " (a int) with ('connector' = 'PRINT')");
+        }
+
+        tEnv.getConfig().set(TableConfigOptions.TABLE_RTAS_CTAS_ATOMICITY_ENABLED, true);
+        String dataDir = tmpDataFolder.getAbsolutePath();
+        String sqlFragment =
+                isCreateOrReplace
+                        ? " create or replace table " + tableName
+                        : " replace table " + tableName;
+        String sql =
+                sqlFragment
+                        + " with ('connector' = 'test-staging', 'data-dir' = '"
+                        + dataDir
+                        + "') as select * from t1";
+        if (!isCreateOrReplace && !createDriverTable) {
+            assertThatThrownBy(() -> tEnv.executeSql(sql))
+                    .isInstanceOf(TableException.class)
+                    .hasMessage(
+                            "The table `default_catalog`.`default_database`.`"
+                                    + tableName
+                                    + "` to be replaced doesn't exist."
+                                    + " You can try to use CREATE TABLE AS statement or CREATE OR REPLACE TABLE AS statement.");
+        } else {
+            tEnv.executeSql(sql).await();
+            if (createDriverTable) {
+                assertThat(tEnv.listTables()).contains(tableName);
+            } else {
+                assertThat(tEnv.listTables()).doesNotContain(tableName);
+            }
+            verifyDataFile(dataDir, "data");
+            assertThat(TestSupportsStagingTableFactory.JOB_STATUS_CHANGE_PROCESS).hasSize(2);
+            assertThat(TestSupportsStagingTableFactory.JOB_STATUS_CHANGE_PROCESS)
+                    .contains("begin", "commit");
+            assertThat(TestSupportsStagingTableFactory.STAGING_PURPOSE_LIST).hasSize(1);
+            if (isCreateOrReplace) {
+                assertThat(TestSupportsStagingTableFactory.STAGING_PURPOSE_LIST)
+                        .contains(SupportsStaging.StagingPurpose.CREATE_OR_REPLACE_TABLE_AS);
+            } else {
+                assertThat(TestSupportsStagingTableFactory.STAGING_PURPOSE_LIST)
+                        .contains(SupportsStaging.StagingPurpose.REPLACE_TABLE_AS);
+            }
+        }
+    }
+
+    @Test
+    void testAtomicReplaceTableAsWithException(@TempDir Path temporaryFolder) throws Exception {
+        commonTestForAtomicReplaceTableAsWithException(
+                "atomic_replace_table_fail", false, temporaryFolder.toFile());
+    }
+
+    @Test
+    void testAtomicCreateOrReplaceTableAsWithException(@TempDir Path temporaryFolder)
+            throws Exception {
+        commonTestForAtomicReplaceTableAsWithException(
+                "atomic_create_or_replace_table_fail", false, temporaryFolder.toFile());
+    }
+
+    private void commonTestForAtomicReplaceTableAsWithException(
+            String tableName, boolean isCreateOrReplace, File tmpDataFolder) {
```

Comment:
get warning from my IDE:
`Actual value of parameter 'isCreateOrReplace' is always 'false' `

Reply:
good catch! It was an oversight on my part.